File: backend/src/s3.js
import { S3Client, PutObjectCommand, GetObjectCommand, CreateBucketCommand, HeadBucketCommand, DeleteObjectCommand } from '@aws-sdk/client-s3';

const endpoint = process.env.MINIO_ENDPOINT || 'localhost';
const port = process.env.MINIO_PORT || '9000';
const accessKey = process.env.MINIO_ACCESS_KEY || 'minioadmin';
const secretKey = process.env.MINIO_SECRET_KEY || 'minioadminpassword';
const useSSL = process.env.MINIO_USE_SSL === 'true';

const s3 = new S3Client({
  endpoint: `http${useSSL ? 's' : ''}://${endpoint}:${port}`,
  credentials: {
    accessKeyId: accessKey,
    secretAccessKey: secretKey
  },
  forcePathStyle: true, // Needed for MinIO compatibility
  region: 'us-east-1'
});

const BUCKET_NAME = 'linkhub-assets';

// Helper to check if bucket exists, create if not
export const initS3 = async () => {
  try {
    await s3.send(new HeadBucketCommand({ Bucket: BUCKET_NAME }));
    console.log(`MinIO bucket "${BUCKET_NAME}" already exists.`);
  } catch (error) {
    // If bucket does not exist, create it
    if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
      try {
        await s3.send(new CreateBucketCommand({ Bucket: BUCKET_NAME }));
        console.log(`Successfully created MinIO bucket "${BUCKET_NAME}".`);
      } catch (createError) {
        console.error('Failed to create MinIO bucket:', createError);
        throw createError;
      }
    } else {
      console.error('Failed checking bucket existence in MinIO:', error);
      throw error;
    }
  }
};

/**
 * Uploads a buffer as a file to MinIO
 * @param {string} key - Unique storage path/key
 * @param {Buffer} buffer - File buffer
 * @param {string} contentType - Mime type of file
 */
export const uploadAsset = async (key, buffer, contentType) => {
  const command = new PutObjectCommand({
    Bucket: BUCKET_NAME,
    Key: key,
    Body: buffer,
    ContentType: contentType
  });
  await s3.send(command);
  return key;
};

/**
 * Downloads a file buffer from MinIO
 * @param {string} key - Unique storage path/key
 * @returns {Promise<Buffer>} The file buffer
 */
export const downloadAsset = async (key) => {
  const command = new GetObjectCommand({
    Bucket: BUCKET_NAME,
    Key: key
  });
  
  const response = await s3.send(command);
  const stream = response.Body;
  
  const chunks = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks);
};

/**
 * Deletes a file from MinIO
 * @param {string} key - Unique storage path/key
 */
export const deleteAsset = async (key) => {
  const command = new DeleteObjectCommand({
    Bucket: BUCKET_NAME,
    Key: key
  });
  await s3.send(command);
};

export default {
  initS3,
  uploadAsset,
  downloadAsset,
  deleteAsset
};

File: backend/src/s3/s3.go
package s3

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

var (
	Client     *minio.Client
	BucketName = "linkhub-assets"
)

// InitS3 connects to MinIO and ensures the target bucket exists.
func InitS3() error {
	endpoint := os.Getenv("MINIO_ENDPOINT")
	if endpoint == "" {
		endpoint = "localhost"
	}
	port := os.Getenv("MINIO_PORT")
	if port == "" {
		port = "9000"
	}
	accessKey := os.Getenv("MINIO_ACCESS_KEY")
	if accessKey == "" {
		accessKey = "minioadmin"
	}
	secretKey := os.Getenv("MINIO_SECRET_KEY")
	if secretKey == "" {
		secretKey = "minioadminpassword"
	}
	useSSLStr := os.Getenv("MINIO_USE_SSL")
	useSSL, _ := strconv.ParseBool(useSSLStr)

	fullEndpoint := fmt.Sprintf("%s:%s", endpoint, port)

	var err error
	Client, err = minio.New(fullEndpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
		Secure: useSSL,
	})
	if err != nil {
		return fmt.Errorf("failed to initialize MinIO client: %w", err)
	}

	ctx := context.Background()
	exists, err := Client.BucketExists(ctx, BucketName)
	if err != nil {
		return fmt.Errorf("failed to check if bucket exists: %w", err)
	}

	if !exists {
		err = Client.MakeBucket(ctx, BucketName, minio.MakeBucketOptions{Region: "us-east-1"})
		if err != nil {
			return fmt.Errorf("failed to create bucket: %w", err)
		}
		log.Printf("Successfully created MinIO bucket \"%s\".\n", BucketName)
	} else {
		log.Printf("MinIO bucket \"%s\" already exists.\n", BucketName)
	}

	return nil
}

// UploadAsset uploads a byte slice to MinIO.
func UploadAsset(key string, data []byte, contentType string) (string, error) {
	ctx := context.Background()
	reader := bytes.NewReader(data)
	_, err := Client.PutObject(ctx, BucketName, key, reader, int64(len(data)), minio.PutObjectOptions{
		ContentType: contentType,
	})
	if err != nil {
		return "", fmt.Errorf("failed to upload object: %w", err)
	}
	return key, nil
}

// DownloadAsset downloads an object from MinIO and returns it as a byte slice.
func DownloadAsset(key string) ([]byte, error) {
	ctx := context.Background()
	object, err := Client.GetObject(ctx, BucketName, key, minio.GetObjectOptions{})
	if err != nil {
		return nil, fmt.Errorf("failed to download object: %w", err)
	}
	defer object.Close()

	buf := new(bytes.Buffer)
	if _, err := io.Copy(buf, object); err != nil {
		return nil, fmt.Errorf("failed to read object payload: %w", err)
	}

	return buf.Bytes(), nil
}

// DeleteAsset deletes an object from MinIO.
func DeleteAsset(key string) error {
	ctx := context.Background()
	err := Client.RemoveObject(ctx, BucketName, key, minio.RemoveObjectOptions{})
	if err != nil {
		return fmt.Errorf("failed to delete object: %w", err)
	}
	return nil
}
